File: app/src/main/java/com/siju/acexplorer/main/model/helper/RootHelper.java
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.siju.acexplorer.main.model.helper;

import android.util.Log;

import com.siju.acexplorer.common.types.FileInfo;
import com.siju.acexplorer.logging.Logger;
import com.siju.acexplorer.main.model.data.FileDataFetcher;
import com.stericson.RootShell.exceptions.RootDeniedException;
import com.stericson.RootShell.execution.Command;
import com.stericson.RootTools.RootTools;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Locale;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;

import static com.siju.acexplorer.main.model.groups.Category.FILES;

public class RootHelper {

    private static final String TAG = "RootHelper";

    public static synchronized ArrayList<String> executeCommand(String cmd) {
        Log.e(TAG, "executeCommand: "+cmd);
        final ArrayList<String> list = new ArrayList<>();
//        final CountDownLatch countDownLatch = new CountDownLatch(1);
        final AtomicReference<ArrayList<String>> resultRef = new AtomicReference<>();
        Command command = new Command(0, cmd) {
            @Override
            public void commandOutput(int id, String line) {
                super.commandOutput(id, line);
                list.add(line);
                Log.e(TAG, "command commandOutput:"+line);
            }

            @Override
            public void commandTerminated(int id, String reason) {
                super.commandTerminated(id, reason);
                Log.e(TAG, "command terminated:"+reason);
            }

            @Override
            public void commandCompleted(int id, int exitcode) {
                super.commandCompleted(id, exitcode);
                Log.e(TAG, "command commandCompleted:"+list.size());
                resultRef.set(list);
//                countDownLatch.countDown();
            }
        };
        try {
            RootTools.getShell(true).add(command);
            waitForCommand(command, -1);
//            countDownLatch.await();
        } catch (IOException | RootDeniedException | TimeoutException e) {
            e.printStackTrace();
        }
        return resultRef.get();
    }

    public static void runAndWait(String cmd) {

        Command c = new Command(0, cmd) {
            @Override
            public void commandOutput(int i, String s) {
                Logger.log(TAG, "commandOutput: i=" + i + " s=" + s);

            }

            @Override
            public void commandTerminated(int i, String s) {
                Logger.log(TAG, "commandTerminated: i=" + i + " s=" + s);
            }

            @Override
            public void commandCompleted(int i, int i2) {
                Logger.log(TAG, "commandCompleted: i=" + i + " i2=" + i2);
            }
        };
        try {
            RootTools.getShell(true).add(c);
        } catch (Exception e) {
            return;
        }
        waitForCommand(c, -1);
    }

    private static boolean waitForCommand(Command cmd, long time) {
        long t = 0;
        while (!cmd.isFinished()) {
            //noinspection SynchronizationOnLocalVariableOrMethodParameter
            synchronized (cmd) {
                try {
                    if (!cmd.isFinished()) {
                        cmd.wait(2000);
                        t += 2000;
                        if (t != -1 && t >= time) {
                            return true;
                        }

                    }
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            if (!cmd.isExecuting() && !cmd.isFinished()) {
                return false;
            }
        }

        Logger.log(TAG, "Command Finished!"+cmd);
        return true;
    }

    public static String getCommandLineString(String input) {
        return input.replaceAll(UNIX_ESCAPE_EXPRESSION, "\\\\$1");
    }

    private static final String UNIX_ESCAPE_EXPRESSION = "(\\(|\\)|\\[|\\]|\\s|\'|\"|`|\\{|\\}|&|\\\\|\\?)";


    public static String parseFilePermission(File f) {
        String per = "";
        if (f.canRead()) {
            per = per + "r";
        }
        if (f.canWrite()) {
            per = per + "w";
        }
        if (f.canExecute()) {
            per = per + "x";
        }
        return per;
    }


    public static ArrayList<FileInfo> getRootedList(String path, boolean root,
                                                    boolean showHidden) {
        ArrayList<FileInfo> fileInfoArrayList = new ArrayList<>();

        String hidden = " ";
        if (showHidden) {
            hidden = "a ";
        }
        ArrayList<String> list;
        long time = System.currentTimeMillis();
        Log.e(TAG, "getRootedList: time:" + time);
        boolean rootAccessGiven = RootTools.isAccessGiven();
        boolean rooted = root || rootAccessGiven;
        if (rooted) {
            list = executeCommand("ls -l " + getCommandLineString(path));
            long newTime = System.currentTimeMillis();
            Log.e(TAG, "getRootedList: time taken for ls:" + (newTime - time) +" list:"+list);
            if (list != null) {
                for (int i = 0; i < list.size(); i++) {
                    String file1 = list.get(i);
                    parseFileNew(path, file1, fileInfoArrayList);
                }
            }
        }
        return fileInfoArrayList;
    }

    public static boolean fileExists(String path) {
        File f = new File(path);
        String p = f.getParent();
        if (p != null && p.length() > 0) {
            ArrayList<FileInfo> ls = FileDataFetcher.Companion.getFilesList(p, false, true, false);
            for (FileInfo strings : ls) {
                if (strings.getFilePath() != null && strings.getFilePath().equals(path)) {
                    return true;
                }
            }
        }
        return false;
    }


    private static SimpleDateFormat simpledateformat = new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.ENGLISH);

    private static void parseFileNew(String path, String result, ArrayList<FileInfo> fileInfoArrayList) {
        String[] array = result.trim().split("\\s+");
        int arrayLength = array.length;
        Log.e(TAG, "parseFileNew: arrayLength:" + arrayLength);
        if (array.length > 3) {
            String trimName;
            String date;
            String permission = array[0];
            String name;
            long size = 0L;

            boolean isDirectory = permission.startsWith("d");
            boolean isLink = permission.startsWith("l");

            if (isDirectory) {
                trimName = array[arrayLength - 1];
                name = trimName;
                date = array[arrayLength - 3] + " " + array[arrayLength - 2];
            } else if (isLink) {
                trimName = array[arrayLength - 3];
                name = trimName;
                date = array[arrayLength - 5] + " " + array[arrayLength - 4];
            } else {
                trimName = array[arrayLength - 1];
                name = trimName;
                size = Long.valueOf(array[arrayLength - 4]);
                date = array[arrayLength - 3] + " " + array[arrayLength - 2];
            }
            long lastModified = getTimeinMillis(date);
            String filePath = fixSlashes(path + File.separator + name);
            fileInfoArrayList.add(new FileInfo(FILES, name, filePath,
                    lastModified, size, isDirectory, null,
                    permission, true));
        }
    }

    private static long getTimeinMillis(String date) {
        long timeInMillis = 0;
        try {
            timeInMillis = simpledateformat.parse(date).getTime();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return timeInMillis;
    }

    // Removes duplicate adjacent slashes and any trailing slash.
    private static String fixSlashes(String origPath) {
        boolean lastWasSlash = false;
        char[] newPath = origPath.toCharArray();
        int length = newPath.length;
        int newLength = 0;
        for (int i = 0; i < length; ++i) {
            char ch = newPath[i];
            if (ch == '/') {
                if (!lastWasSlash) {
                    newPath[newLength++] = File.separatorChar;
                    lastWasSlash = true;
                }
            } else {
                newPath[newLength++] = ch;
                lastWasSlash = false;
            }
        }
        // Remove any trailing slash (unless this is the root of the file system).
        if (lastWasSlash && newLength > 1) {
            newLength--;
        }
        return (newLength != length) ? new String(newPath, 0, newLength) : origPath;
    }

}

File: app/src/main/java/com/siju/acexplorer/main/model/helper/RootHelper.kt
/*
 * Copyright (C) 2017 Ace Explorer owned by Siju Sakaria
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
@file:Suppress("RegExpRedundantEscape")
package com.siju.acexplorer.main.model.helper

import android.annotation.SuppressLint
import android.util.Log
import com.siju.acexplorer.common.types.FileInfo
import com.siju.acexplorer.logging.Logger.log
import com.siju.acexplorer.main.model.data.FileDataFetcher.Companion.getFilesList
import com.siju.acexplorer.main.model.groups.Category
import com.stericson.RootShell.exceptions.RootDeniedException
import com.stericson.RootShell.execution.Command
import com.stericson.RootTools.RootTools
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeoutException
import java.util.concurrent.atomic.AtomicReference

private const val TAG = "RootHelper"
private const val COMMAND_WAIT_MS = 2000
private const val UNIX_ESCAPE_EXPRESSION = "(\\(|\\)|\\[|\\]|\\s|\'|\"|`|\\{|\\}|&|\\\\|\\?)"

@SuppressLint("StaticFieldLeak")
object RootHelper {
    private val simpleDateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.ENGLISH)

    @Synchronized
    fun executeCommand(cmd: String): ArrayList<String> {
        Log.d(TAG, "executeCommand: $cmd")
        val list = ArrayList<String>()
        //        final CountDownLatch countDownLatch = new CountDownLatch(1);
        val resultRef = AtomicReference<ArrayList<String>>()
        val command: Command = object : Command(0, cmd) {
            override fun commandOutput(id: Int, line: String) {
                super.commandOutput(id, line)
                list.add(line)
                Log.d(TAG, "command commandOutput:$line")
            }

            override fun commandTerminated(id: Int, reason: String) {
                super.commandTerminated(id, reason)
                Log.d(TAG, "command terminated:$reason")
            }

            override fun commandCompleted(id: Int, exitcode: Int) {
                super.commandCompleted(id, exitcode)
                Log.d(TAG, "command commandCompleted:" + list.size)
                resultRef.set(list)
                //                countDownLatch.countDown();
            }
        }
        try {
            RootTools.getShell(true).add(command)
            waitForCommand(command,1)
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: RootDeniedException) {
            e.printStackTrace()
        } catch (e: TimeoutException) {
            e.printStackTrace()
        }
        return resultRef.get()
    }

    fun runAndWait(cmd: String?) {
        val c: Command = object : Command(0, cmd) {
            override fun commandOutput(i: Int, s: String) {
                log(TAG, "commandOutput: i=$i s=$s")
            }

            override fun commandTerminated(i: Int, s: String) {
                log(TAG, "commandTerminated: i=$i s=$s")
            }

            override fun commandCompleted(i: Int, i2: Int) {
                log(TAG, "commandCompleted: i=$i i2=$i2")
            }
        }
        try {
            RootTools.getShell(true).add(c)
        } catch (e: Exception) {
            return
        }
        waitForCommand(c, -1)
    }

    private fun waitForCommand(command: Command, time: Long = -1L): Boolean {
        var t: Long = 0
        while (!command.isFinished) {
            synchronized(command) {
                try {
                    if (!command.isFinished) {
                        t += COMMAND_WAIT_MS
                        if (t != -1L && t >= time) {
                            return true
                        }
                    }
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
            if (!command.isExecuting && !command.isFinished) {
                return false
            }
        }
        log(TAG, "Command Finished!$command")
        return true
    }

    fun getCommandLineString(input: String): String {
        return input.replace(UNIX_ESCAPE_EXPRESSION.toRegex(), "\\\\$1")
    }

    fun parseFilePermission(f: File): String {
        var per = ""
        if (f.canRead()) {
            per += "r"
        }
        if (f.canWrite()) {
            per += "w"
        }
        if (f.canExecute()) {
            per += "x"
        }
        return per
    }

    fun getRootedList(path: String, root: Boolean,
                      showHidden: Boolean): ArrayList<FileInfo> {
        val fileInfoArrayList = ArrayList<FileInfo>()
        var hidden = " "
        if (showHidden) {
            hidden = "a "
        }
        val list: ArrayList<String>
        val time = System.currentTimeMillis()
        Log.d(TAG, "getRootedList: time:$time")
        val rootAccessGiven = RootTools.isAccessGiven()
        val rooted = root || rootAccessGiven
        if (rooted) {
            list = executeCommand("ls -l" + hidden + getCommandLineString(path))
            val newTime = System.currentTimeMillis()
            Log.d(TAG, "getRootedList: time taken for ls:" + (newTime - time) + " list:" + list)
            for (i in list.indices) {
                val file1 = list[i]
                parseFileNew(path, file1, fileInfoArrayList)
            }
        }
        return fileInfoArrayList
    }

    fun fileExists(path: String?): Boolean {
        path ?: return true
        val file = File(path)
        val parent = file.parent
        if (parent != null && parent.isNotEmpty()) {
            val ls = getFilesList(parent, root = false, showHidden = true, ringtonePicker = false)
            for (strings in ls) {
                if (strings.filePath != null && strings.filePath == path) {
                    return true
                }
            }
        }
        return false
    }

    private fun parseFileNew(path: String, result: String, fileInfoArrayList: ArrayList<FileInfo>) {
        val array = result.trim { it <= ' ' }.split("\\s+").toTypedArray()
        val arrayLength = array.size
        Log.d(TAG, "parseFileNew: arrayLength:$arrayLength")
        if (array.size > 3) {
            val trimName: String
            val date: String
            val permission = array[0]
            val name: String
            var size = 0L
            val isDirectory = permission.startsWith("d")
            val isLink = permission.startsWith("l")
            when {
                isDirectory -> {
                    trimName = array[arrayLength - 1]
                    name = trimName
                    date = array[arrayLength - 3] + " " + array[arrayLength - 2]
                }
                isLink -> {
                    trimName = array[arrayLength - 3]
                    name = trimName
                    date = array[arrayLength - 5] + " " + array[arrayLength - 4]
                }
                else -> {
                    trimName = array[arrayLength - 1]
                    name = trimName
                    size = java.lang.Long.valueOf(array[arrayLength - 4])
                    date = array[arrayLength - 3] + " " + array[arrayLength - 2]
                }
            }
            val lastModified = getTimeInMillis(date)
            val filePath = fixSlashes(path + File.separator + name)
            fileInfoArrayList.add(FileInfo(Category.FILES, name, filePath,
                    lastModified, size, isDirectory, null,
                    permission, true))
        }
    }

    private fun getTimeInMillis(date: String): Long {
        var timeInMillis = 0L
        try {
            val dateTime = simpleDateFormat.parse(date)
            dateTime?.let {
                timeInMillis = it.time
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return timeInMillis
    }

    // Removes duplicate adjacent slashes and any trailing slash.
    private fun fixSlashes(origPath: String): String {
        var lastWasSlash = false
        val newPath = origPath.toCharArray()
        val length = newPath.size
        var newLength = 0
        for (i in 0 until length) {
            val ch = newPath[i]
            if (ch == '/') {
                if (!lastWasSlash) {
                    newPath[newLength++] = File.separatorChar
                    lastWasSlash = true
                }
            } else {
                newPath[newLength++] = ch
                lastWasSlash = false
            }
        }
        // Remove any trailing slash (unless this is the root of the file system).
        if (lastWasSlash && newLength > 1) {
            newLength--
        }
        return if (newLength != length) String(newPath, 0, newLength) else origPath
    }
}